# About
This notebook encodes the various columns of metadata.csv so that they can be used for a machine learning model.
It creates a new dataframe of the encoded values, drops columns that are redundant, and exports to md_encoded.csv.

Author: Riley Denn

In [1]:
import pandas as pd
import numpy as np
import os
import json
from sklearn.preprocessing import MinMaxScaler, LabelEncoder

In [8]:
with open('../../config.json', 'r') as config_file:
    config = json.load(config_file)

print(config['drive_path'])

DRIVEPATH = config['drive_path']
PATH = DRIVEPATH + "/[EXTERNAL] breakthrough_tech_ai_f24/data"

/Users/rileydenn/Library/CloudStorage/GoogleDrive-rileydenn@gmail.com/.shortcut-targets-by-id/1FsOPywSgK_wZmrVrSTBVi4q8G3Mg_yMJ/Team-Fermata-Energy


In [9]:
md_full = pd.read_csv(PATH + "/metadata.csv")

FOLDER_PATH = PATH + "/building_data"

existing_ids = set()

for folder_name in os.listdir(FOLDER_PATH):
    try:
        existing_ids.add(int(folder_name))
    except ValueError:
        continue

md = md_full[md_full['bldg_id'].isin(existing_ids)]

In [10]:
md.shape

(32000, 15)

In [11]:
md.columns

Index(['bldg_id', 'in.building_america_climate_zone', 'in.iso_rto_region',
       'in.state', 'in.cluster_id', 'in.cluster_name', 'in.heating_fuel',
       'in.interior_lighting_generation', 'in.floor_area_category',
       'in.service_water_heating_fuel', 'in.comstock_building_type',
       'in.comstock_building_type_group', 'in.vintage', 'in.sqft',
       'in.nhgis_county_gisjoin'],
      dtype='object')

In [12]:
md.head()

bldg_id in.building_america_climate_zone in.iso_rto_region in.state  \
410     48709                          Hot-Dry               NaN       CA   
494     58658                             Cold               NaN       CO   
905    105885                        Hot-Humid               NaN       HI   
1036   119634                             Cold               PJM       IL   
1757   205046                      Mixed-Humid             NYISO       NY   

      in.cluster_id                            in.cluster_name  \
410           109.0                          CA Climate Zone 9   
494            36.0  the Cheyenne-Denver-Colorado Springs Area   
905            42.0                                     Hawaii   
1036           82.0  the Chicago to Green Bay and Madison Area   
1757           54.0             the Greater New York City Area   

      in.heating_fuel in.interior_lighting_generation in.floor_area_category  \
410       Electricity                        gen4_led          200001_500000   
494        NaturalGas                 gen2_t8_halogen          100001_200000   
905        NaturalGas           gen1_t12_incandescent            500001_1mil   
1036       NaturalGas                 gen2_t8_halogen            500001_1mil   
1757  DistrictHeating                 gen2_t8_halogen          200001_500000   

     in.service_water_heating_fuel in.comstock_building_type  \
410                    Electricity               LargeOffice   
494                    Electricity               LargeOffice   
905                     NaturalGas                LargeHotel   
1036                    NaturalGas                LargeHotel   
1757               DistrictHeating               LargeOffice   

     in.comstock_building_type_group    in.vintage   in.sqft  \
410                           Office  1960 to 1969  350000.0   
494                           Office   Before 1946  150000.0   
905                          Lodging  1970 to 1979  750000.0   
1036                         Lodging  1990 to 1999  750000.0   
1757                          Office   Before 1946  350000.0   

     in.nhgis_county_gisjoin  
410                 G0600370  
494                 G0800310  
905                 G1500030  
1036                G1700310  
1757                G3600610

In [13]:
md.dtypes

bldg_id                               int64
in.building_america_climate_zone     object
in.iso_rto_region                    object
in.state                             object
in.cluster_id                       float64
in.cluster_name                      object
in.heating_fuel                      object
in.interior_lighting_generation      object
in.floor_area_category               object
in.service_water_heating_fuel        object
in.comstock_building_type            object
in.comstock_building_type_group      object
in.vintage                           object
in.sqft                             float64
in.nhgis_county_gisjoin              object
dtype: object

In [14]:
md.isnull().any()

bldg_id                             False
in.building_america_climate_zone    False
in.iso_rto_region                    True
in.state                            False
in.cluster_id                       False
in.cluster_name                     False
in.heating_fuel                     False
in.interior_lighting_generation     False
in.floor_area_category              False
in.service_water_heating_fuel       False
in.comstock_building_type           False
in.comstock_building_type_group     False
in.vintage                          False
in.sqft                             False
in.nhgis_county_gisjoin             False
dtype: bool

In [15]:
for col in md.columns:
    print(f"{col}: {md[col].unique()}\n"
    f"number of unique values: {np.size(md[col].unique())}\n")

bldg_id: [ 48709  58658 105885 ... 332085 332104 332130]
number of unique values: 32000

in.building_america_climate_zone: ['Hot-Dry' 'Cold' 'Hot-Humid' 'Mixed-Humid' 'Very Cold' 'Subarctic'
 'Mixed-Dry' 'Marine']
number of unique values: 8

in.iso_rto_region: [nan 'PJM' 'NYISO' 'ERCOT' 'MISO' 'SPP' 'CAISO' 'NEISO']
number of unique values: 8

in.state: ['CA' 'CO' 'HI' 'IL' 'NY' 'PA' 'TX' 'AL' 'AK' 'AZ' 'AR' 'DE' 'DC' 'FL'
 'GA' 'ID' 'IN' 'IA' 'KS' 'KY' 'LA' 'ME' 'MD' 'MI' 'MN' 'MS' 'MO' 'MT'
 'NE' 'NV' 'NH' 'NJ' 'NM' 'NC' 'ND' 'OH' 'OK' 'OR' 'SC' 'SD' 'TN' 'UT'
 'VT' 'VA']
number of unique values: 44

in.cluster_id: [109.  36.  42.  82.  54.  63.  74.  75.  53.  85.  81.  76.  37.  71.
  24.  31.   1.  60.  64.  33.  27. 104. 101. 108. 105. 110. 113. 111.
 112. 114.  86.  57.  12.   8.  49.  39.  55.  46.  84.  50.  18.  35.
   9.  26.  66.  68.  17.  47.   3.  51.   4.  56.  78.   0.  43.  30.
  40.  62.  70.  61.  59.  11.  16.  23.  72.  21.  77.  41.  20.  14.
  80.   6.  19.  38.

In [16]:
md.columns

Index(['bldg_id', 'in.building_america_climate_zone', 'in.iso_rto_region',
       'in.state', 'in.cluster_id', 'in.cluster_name', 'in.heating_fuel',
       'in.interior_lighting_generation', 'in.floor_area_category',
       'in.service_water_heating_fuel', 'in.comstock_building_type',
       'in.comstock_building_type_group', 'in.vintage', 'in.sqft',
       'in.nhgis_county_gisjoin'],
      dtype='object')

In [17]:
# Percentage null values for in.iso_rto_region
md["in.iso_rto_region"].isnull().sum() / len(md) * 100

37.30625

In [18]:
md.loc[:, "in.iso_rto_region"] = md["in.iso_rto_region"].fillna("none")

In [19]:
md_encoded = pd.DataFrame()

md_encoded = pd.get_dummies(md, columns=[
    'in.building_america_climate_zone',
    'in.iso_rto_region',
    'in.heating_fuel', 
    'in.interior_lighting_generation',
    'in.service_water_heating_fuel',
    'in.comstock_building_type',
    'in.comstock_building_type_group'
])

# Manually mapping ordinal encoding for year ranges
vintage_mapping = {
    'Before 1946': 0,
    '1946 to 1959': 1,
    '1960 to 1969': 2,
    '1970 to 1979': 3,
    '1980 to 1989': 4,
    '1990 to 1999': 5,
    '2000 to 2012': 6,
    '2013 to 2018': 7
}
md_encoded['in.vintage'] = md['in.vintage'].map(vintage_mapping)

# Label encoding for states
label_encoder = LabelEncoder()
md_encoded["in.state"] = label_encoder.fit_transform(md["in.state"])

# Choose to normalize or not after choosing model
# Scale/normalize sqft values to range between 0 and 1
# scaler = MinMaxScaler()
# md_encoded["in.sqft"] = scaler.fit_transform(md_encoded[["in.sqft"]])

# Drop unwanted columns
columns_to_drop = [
    'in.cluster_name',
    'in.floor_area_category',
    'in.nhgis_county_gisjoin'
]
md_encoded = md_encoded.drop(columns=columns_to_drop)

In [20]:
md_encoded

bldg_id  in.state  in.cluster_id  in.vintage   in.sqft  \
410       48709         4          109.0           2  350000.0   
494       58658         5           36.0           0  150000.0   
905      105885        10           42.0           3  750000.0   
1036     119634        13           82.0           5  750000.0   
1757     205046        32           54.0           0  350000.0   
...         ...       ...            ...         ...       ...   
328562   332060        42           39.0           1    7500.0   
328585   332083        42           22.0           2   17500.0   
328587   332085        42           11.0           0   17500.0   
328606   332104        42           22.0           3   17500.0   
328632   332130        42           22.0           4   17500.0   

        in.building_america_climate_zone_Cold  \
410                                     False   
494                                      True   
905                                     False   
1036                                     True   
1757                                    False   
...                                       ...   
328562                                  False   
328585                                  False   
328587                                  False   
328606                                  False   
328632                                  False   

        in.building_america_climate_zone_Hot-Dry  \
410                                         True   
494                                        False   
905                                        False   
1036                                       False   
1757                                       False   
...                                          ...   
328562                                     False   
328585                                     False   
328587                                     False   
328606                                     False   
328632                                     False   

        in.building_america_climate_zone_Hot-Humid  \
410                                          False   
494                                          False   
905                                           True   
1036                                         False   
1757                                         False   
...                                            ...   
328562                                       False   
328585                                       False   
328587                                       False   
328606                                       False   
328632                                       False   

        in.building_america_climate_zone_Marine  \
410                                       False   
494                                       False   
905                                       False   
1036                                      False   
1757                                      False   
...                                         ...   
328562                                    False   
328585                                    False   
328587                                    False   
328606                                    False   
328632                                    False   

        in.building_america_climate_zone_Mixed-Dry  ...  \
410                                          False  ...   
494                                          False  ...   
905                                          False  ...   
1036                                         False  ...   
1757                                         False  ...   
...                                            ...  ...   
328562                                       False  ...   
328585                                       False  ...   
328587                                       False  ...   
328606                                       False  ...   
328632                                       False  ...   

        in.comstock_building_t

In [21]:
md_encoded.to_csv('md_one_hot_encoded.csv', index=False)

# Considerations:
- md["in.state"] was encoded with label encoding, assigning a number value to each state. Could consider target encoding instead, replacing state column with the average energy load for buildings in each state.    

- md["in.nhgis_county_gisjoin"] - represents coordinates, used for mapping weather data to spatial coordinates. Could use plotly or geographic plotly for visualization. Can be dropped for model dev.    

- md["in.iso_rto_region"] has ~37% nulls. Changed nans to the string "none" then one hot encoded. ISO RTO reigons refer to the governing agency and is reigonal, so has correlation to other geographic features. May not be needed for model but is included in md_encoded just in case.

- md["in.comstock_building_type"] is similar to the column md["in.comstock_building_type_group"], which is a simplified version of this feature with fewer unique values, while still capturing the various building types at a broad level. Both have been included in md_encoded. For modeling, experiment with using both or one or the other.

- md_encoded["in.sqft"] left as is - can choose to normalize sqft or not after choosing model.  
Code to normalize sqft values to range between 0 and 1:
```python
scaler = MinMaxScaler()  
md_encoded["in.sqft"] = scaler.fit_transform(md_encoded[["in.sqft"]])